### ISSUE TITLE
Update-DbaInstance not installing latest security update for a CU

### ISSUE THREAD

**Issue by dokier** [dataplat/dbatools#9327]
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

WARNING: [10:28:44][Get-SqlInstanceUpdate] 14.0.3460 on computer [mysqlserver.domain.com] is already  the latest available.                                                                                         

### Steps to Reproduce

I have a SQL Server 2017 instance I was trying to patch to the latest available build version (14.0.3465.1) This is [5029376 Security update for SQL Server 2017 CU31: October 10, 2023](https://support.microsoft.com/en-us/help/5029376). You can also refer to https://sqlserverbuilds.blogspot.com/ 

```powershell
# provide your command(s) executed pertaining to dbatools
Update-DbaInstance -ComputerName mysqlserver -Restart -Path \\share\Updates -Confirm:$false
# please include variable values (redacted or fake if needed) for reference
```

### Please confirm that you are running the most recent version of dbatools

2.1.14

### Other details or mentions

dbatools-buildref-index.json has the latest information

`        {
            "CU": "CU31",
            "Version": "14.0.3456",
            "KBList": "5016884"
        },
        {
            "Version": "14.0.3460",
            "KBList": "5021126"
        },
        {
            "Version": "14.0.3465",
            "KBList": "5029376"
        },`

but because the last two elements do not have "CU" tags, these are not installed.  This all comes down to Get-SqlInstanceUpdate calling Test-DbaBuild on line 109.

`$latestCU = Test-DbaBuild -Build $currentVersion.BuildLevel -MaxBehind '0CU'`

If this is by designed, should we have an option in Update-DbaInstance to override this so that it includes security updates for a CU? You can always specify the KB in the cmdlet but that doesn't scale well when you're passing a long list of servers that includes different versions of SQL.

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe)

### PowerShell Host Version

Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.1.20348.2227                                                                          
PSEdition                      Desktop                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.20348.2227                                                                         
CLRVersion                     4.0.30319.42000                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 

### SQL Server Edition and Build number

SQL 2017, (14.0.3465.1) now, but that's because I ended up using Update-DbaInstance by specifying the KB number. 

### .NET Framework Version

PSChildName Version
----------- -------
Client      4.8.04161
Full        4.8.04161
Client      4.0.0.0

**Comment by MichelLalancetteNortera** [comment]
Currently experiencing the same issue. I tried to apply a Security update to 4 different servers using SQL 2017 and SQL 2019.

**Comment by dokier** [comment]
@niphlod You could, but you’d need additional logic to filter and pass only the servers that require that specific KB. The cmdlet itself includes the logic to install the latest cumulative update against a list of servers in this manner; however, it doesn’t account for security updates (CU+GDR). Microsoft considers these CU + GDR not to be confused with RTM + GDR updates, which are two different patching paths.

https://learn.microsoft.com/en-us/troubleshoot/sql/releases/download-and-install-latest-updates

`Update-DbaInstance -ComputerName $ServerList -Restart -Path \\share\Updates -Confirm:$false`

**Comment by niphlod** [comment]
Hum. There's a confusione here (I think). RTM+GDR vs CU+GDR has nothing to
do with Update-DbaInstance. The function provides two sets of parameters:
-Kb as in "apply this KB to this instance" and "servicelevel" as in the
Legacy SP or the "modern" CU (with -maxbehind, too, for the laziest of lazy
dbas).

So, we have "maximum precision" vs "sane defaults".

It has nothing, nor ever had, "install whatever is the latest, regardless
of that KB being not marked as a SP or a CU", which seems to be what you
want: please consider that that "latest" could very well be an hotfix
nobody needs except for someone who is experiencing particular issues.

Leaving SPs aside, point here is that not every KB has the same "confidence
level" a CU has, and mindlessly applying the latest KB could result in
disruptions (read: "in-sane").

With CUs, there's at LEAST a minimum confidence that that KB won't wreak
havoc (others, that are NOT CUs, aren't seemigly tested that much, or
they'd come with a "CU" flag themselves).

So, point remains about the race between the DBA (or someone who asked the
DBA) to apply and update with Little confidence ("in-sane") and Microsoft
releasing a new CU, with more confidence ("sane").

Even if someone wanted (like you) the possibility to install every KB that
is not an hotfix but only a security patch, we'd need another property for
each KB to tag KBs that aren't CUs as either "security related" or
"hotfixes", and then filter them on that.

Then there's the added conundrum of CUs being cumulative and hotfixes being
installabile only on top of the CU they're based upon.

It's quite a machinery to build, with no official sources, and the very
little benefit over asking users to pin their "wanted" KB for each SQL
version they manage and pass it as a parameter ("maximum precision"), when
the DBA wants to win the race and not wait for MS to release a new CU
("sane default").

Frankly, if the majority of DBAs were in a situation were applying always
the latest patch was needed, they'd better ask Microsoft to tag CUs more
frequently or live with the thrill of installing a KB that *may* have
consequences.

Il mar 3 dic 2024, 21:57 Daniel Berber ***@***.***> ha
scritto:

**Comment by dokier** [comment]
@niphlod That sounds good, but perhaps the documentation for the cmdlet should be updated. The second example explicitly states, "Updates all applicable SQL Server installations on SQL1 and SQL2 with the most recent patch," which is also reflected in the GitHub code example. That example just passes the network share and nothing elese.

https://docs.dbatools.io/Update-DbaInstance.html
https://github.com/dataplat/dbatools/blob/master/public/Update-DbaInstance.ps1

**Comment by MichelLalancetteNortera** [comment]
@niphlod While I understand what you're saying, SQL 2017 will not be getting any CUs anymore but I'm still obligated by the IT dept. (for insurance reasons) to apply the latest Security Update within 45 days of release, same thing for 2019 and 2022. Do I agree with this? No. Do I have to comply because the company pays me? Yes.

I did workaround it by registering those Security Update KBs in the update list though so it's good enough for me for now.

**Comment by MichelLalancette** [comment]
@niphlod While I understand what you're saying, SQL 2017 will not be getting any CUs anymore but I'm still obligated by the IT dept. (for insurance reasons) to apply the latest Security Update within 45 days of release, same thing for 2019 and 2022. Do I agree with this? No. Do I have to comply because the company pays me? Yes.

I did workaround it by registering those Security Update KBs in the update list though so it's good enough for me for now.

**Comment by niphlod** [comment]
@MichelLalancette  all you'd ideally need to do is: "IT dept forces me to install kb 12345 on 2017?", "Update-DbaInstance -SqlInstance the2017instance -KB 12345".

**Comment by MichelLalancette** [comment]
@niphlod Yep, that's what I do now. At the end of the day, they've been told of the "dangers" of applying patches that quickly so whatever they pay me to do.

**Comment by niphlod** [comment]
or .... some version of this logic
```
$targetBuild = Test-DbaBuild  -Build '14.0' -Latest 
$correctKb = Get-DbaBuild -Build $targetBuild.BuildTarget
Update-DbaInstance .... -KB $correctKb.KbLevel 
```
?

**Comment by niphlod** [comment]
this can be closed, feel free to reopen in case we need to revisit
